lib/mhfile.pl

```diff
@@ -123,225 +124,229 @@ sub file_exists {
-			 (O_WRONLY|O_EXCL|O_CREAT), 0600)) {
-		die qq/ERROR: Unable to create temp file "$tmpfile": $!\n/;
-	    }
-	    last MKTEMP;
-	}
-
-	# Use File::Temp when at all possible
-	if ($_have_File_Temp) {
-	    my @tf_opts = ('DIR' => $dir, 'UNLINK' => 0);
-	    push(@tf_opts, 'SUFFIX' => $suffix)  if $suffix;
-	    ($handle, $tmpfile) = File::Temp::tempfile($template, @tf_opts);
-	    last MKTEMP;
-	}
-
-	$handle = gensym;
-	my($i);
-	for ($i=0; $i < TEMP_MAX_TRIES; ++$i) {
-	    ($tmpfile = $template) =~
-		s/X/$TEMP_CHARS[int(rand($#TEMP_CHARS))]/ge;
-	    $tmpfile  = join($DIRSEP, $dir, $tmpfile);
-	    $tmpfile .= $suffix  if defined($suffix);
-	    last  if sysopen($handle, $tmpfile,
-			     (O_WRONLY|O_EXCL|O_CREAT), 0600);
-	}
-	if ($i >= TEMP_MAX_TRIES) {
-	    die qq/ERROR: Unable to create temp file "$tmpfile": $!\n/;
-	}
+    my $template = shift;               # Filename template
+    my $dir      = shift || $CURDIR;    # Where to write the file
+    my $suffix   = shift;               # Required suffix (optional)
+    my ($handle, $tmpfile);
+
+MKTEMP: {
+        # Do not honor FASTTEMPFILES if a suffix is required, mainly
+        # because things like attachment writing will not work with
+        # FASTTEMPFILES.
+        if ($FastTempFiles && !defined($suffix)) {
+            $handle = gensym;
+            $tmpfile = join($DIRSEP, $dir, $template . $$);
+            if (!sysopen(
+                    $handle, $tmpfile, (O_WRONLY | O_EXCL | O_CREAT), 0600
+                )
+            ) {
+                die qq/ERROR: Unable to create temp file "$tmpfile": $!\n/;
+            }
+            last MKTEMP;
+        }
+
+        # Use File::Temp when at all possible
+        if ($_have_File_Temp) {
+            my @tf_opts = ('DIR' => $dir, 'UNLINK' => 0);
+            push(@tf_opts, 'SUFFIX' => $suffix) if $suffix;
+            ($handle, $tmpfile) = File::Temp::tempfile($template, @tf_opts);
+            last MKTEMP;
+        }
+
+        $handle = gensym;
+        my ($i);
+        for ($i = 0; $i < TEMP_MAX_TRIES; ++$i) {
+            ($tmpfile = $template) =~
+                s/X/$TEMP_CHARS[int(rand($#TEMP_CHARS))]/ge;
+            $tmpfile = join($DIRSEP, $dir, $tmpfile);
+            $tmpfile .= $suffix if defined($suffix);
+            last
+                if sysopen($handle, $tmpfile,
+                (O_WRONLY | O_EXCL | O_CREAT), 0600);
+        }
+        if ($i >= TEMP_MAX_TRIES) {
+            die qq/ERROR: Unable to create temp file "$tmpfile": $!\n/;
+        }
     }
     ($handle, $tmpfile);
 }
 
 sub file_chmod {
-    my $file  = shift;
-    my $perm  = shift || $FilePermsOct;
+    my $file = shift;
+    my $perm = shift || $FilePermsOct;
     ## Capture any die's in case chmod not supported.
     eval {
-	if (chmod(($perm &~ umask), $file) < 1) {
-	    warn qq/Warning: Unable to change "$file" permissions to "/,
-		 sprintf('%o'. $perm),
-		 qq/": $!\n/;
-	}
+        if (chmod(($perm & ~umask), $file) < 1) {
```

Comment:
I don’t know this syntax: `&~`. Does it change something to use `chmod(($perm & ~umask), $file)` instead of `chmod(($perm &~ umask), $file)`? (the `&~ umask` become `& ~umask`)

That is the single change that I’m not confident with (the other changes are straightforward).

Reply:
I agree to you: `&~` looks meaning `& ~`.